Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,133 @@
+# DaemonSet deployment for Prometheus Agent
+
+* Owners:
+  * [haanhvu](https://github.com/haanhvu)
+* Related Tickets:
+  * [#5495](https://github.com/prometheus-operator/prometheus-operator/issues/5495)
+* Other docs:
+  * n/a
+
+This proposal is about designing and implementing the deployment of Prometheus Agent as DaemonSet. Currently, Prometheus Agent can only be deployed as StatefulSet, which could be considered as “cluster-wide” strategy, meaning one or several high-availability Prometheus Agents are responsible for scraping metrics of the whole cluster. Though this works well for many use cases, some use cases may indeed prefer a “node-specific” strategy (DaemonSet), where each Prometheus Agent only scrapes the metrics from the targets located on the same node.
+
+
+## 1. Why
+
+When deploying Prometheus Agent in Kubernetes, three of the biggest users’ concerns are: load distribution, scalability, and security.
+
+DaemonSet deployment solves all these three concerns:
+* Load distribution: Each Prometheus Agent pod will only scrape the targets located on the same node. Even though the targets on some nodes may produce more metrics than other nodes, the load distribution would be reliable enough.
+* Automatic scalability: When new nodes are added to the cluster, new Prometheus Agent pods will be automatically added in the nodes that meet user-defined restrictions (if any).
+* Security: Since the scraped targets are local to the Prometheus Agent pod (on the same node), the scope of security problems is reduced to each node.
+
+DaemonSet deployment is especially more suitable for Prometheus Agent than Prometheus, since the storage requirement for the Agent mode is pretty light compared to a fully functional Prometheus server.
```

Comment:
To be specific: It's currently "only" around 10-20% less memory hungry and does not produce TSDB blocks on disks. Similar CPU wise. I would say that's the main reason for using agent.

More important why agent is better is because trying to query across the nodes, even with Thanos querier on top, is not ideal, because nodes can go up and down and you loose all the data. Agent naturally blocks querying APIs making it easier to scale and prohibits difficult to serve traffic.

```suggestion
DaemonSet deployment is especially more suitable for Prometheus Agent than Prometheus, since agent naturally blocks the querying APIs and forces users to remote write to remote backends. Additionally agent uses (currently) up to 10-20% less memory compared to Prometheus server that produces blocks etc.
```